akka-http-core/src/main/scala/akka/http/impl/engine/ws/WebSocketClientBlueprint.scala

```diff
@@ -59,68 +56,75 @@ object WebSocketClientBlueprint {
-              terminate()
+    class UpgradeStage extends SimpleLinearGraphStage[ByteString] {
+
+      override def createLogic(attributes: Attributes): GraphStageLogic =
+        new GraphStageLogic(shape) with InHandler with OutHandler {
+          // a special version of the parser which only parses one message and then reports the remaining data
+          // if some is available
+          val parser = new HttpResponseParser(settings.parserSettings, HttpHeaderParser(settings.parserSettings)()) {
+            var first = true
+            override def handleInformationalResponses = false
+            override protected def parseMessage(input: ByteString, offset: Int): StateResult = {
+              if (first) {
+                first = false
+                super.parseMessage(input, offset)
+              } else {
+                emit(RemainingBytes(input.drop(offset)))
+                terminate()
+              }
             }
           }
-        }
-        parser.setContextForNextResponse(HttpResponseParser.ResponseContext(HttpMethods.GET, None))
-
-        def onPush(elem: ByteString, ctx: Context[ByteString]): SyncDirective = {
-          parser.parseBytes(elem) match {
-            case NeedMoreData ⇒ ctx.pull()
-            case ResponseStart(status, protocol, headers, entity, close) ⇒
-              val response = HttpResponse(status, headers, protocol = protocol)
-              Handshake.Client.validateResponse(response, subprotocol.toList, key) match {
-                case Right(NegotiatedWebSocketSettings(protocol)) ⇒
-                  result.success(ValidUpgrade(response, protocol))
-
-                  become(transparent)
-                  valve.open()
-
-                  val parseResult = parser.onPull()
-                  require(parseResult == ParserOutput.MessageEnd, s"parseResult should be MessageEnd but was $parseResult")
-                  parser.onPull() match {
-                    case NeedMoreData          ⇒ ctx.pull()
-                    case RemainingBytes(bytes) ⇒ ctx.push(bytes)
-                    case other ⇒
-                      throw new IllegalStateException(s"unexpected element of type ${other.getClass}")
+          parser.setContextForNextResponse(HttpResponseParser.ResponseContext(HttpMethods.GET, None))
+
+          var parseOnPush = true
+
+          override def onPush(): Unit = {
+            val elem = grab(in)
+            if (parseOnPush) {
+              parser.parseBytes(elem) match {
+                case NeedMoreData ⇒ pull(in)
+                case ResponseStart(status, protocol, headers, entity, close) ⇒
+                  val response = HttpResponse(status, headers, protocol = protocol)
+                  Handshake.Client.validateResponse(response, subprotocol.toList, key) match {
+                    case Right(NegotiatedWebSocketSettings(protocol)) ⇒
+                      result.success(ValidUpgrade(response, protocol))
+
+                      parseOnPush = false
```

Comment:
I just tried using `passAlong(in, out)` here, but the tests started throwing errors about the input port being pulled twice. Out of curiosity I tried what I thought should be functionally equivalent:
```scala
setHandler(in, new InHandler {
  override def onPush(): Unit = push(out, grab(in))
})
```
And the tests started passing again without the `parseOnPush` boolean :)

I've never seen `passAlong` until you suggested it, so I'm not sure why it was causing a double-pull.  I left the default argument of `doPull = false`, so it shouldn't have been pulling anything on initialization, right?